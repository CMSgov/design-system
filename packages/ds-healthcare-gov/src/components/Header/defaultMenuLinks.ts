import { VARIATION_NAMES } from './Header';
import localeLink from './localeLink';
import loginLink from './loginLink';
import { translate } from '../i18n';

/**
 * Default menu links for each header variation.
 * Apps can import this method into their app if they need to
 * extend the existing default list of menu links.
 * @param {Function} translate - i18n translator
 * @param {String} locale
 * @param {Boolean} deConsumer
 * @param {String} subpath
 * @param {String} primaryDomain
 * @param {String} switchLocaleLink
 * @param {boolean} hideLoginLink
 * @param {boolean} hideLanguageSwitch
 * @param {boolean} customLinksPassedIn
 * @returns {Object}
 */
export function defaultMenuLinks(
  locale = 'en',
  deConsumer,
  subpath,
  primaryDomain = '',
  switchLocaleLink,
  hideLoginLink,
  hideLogoutLink,
  hideLanguageSwitch,
  customLinksPassedIn
) {
  const isSpanish = locale === 'es';
  const ffmLocalePath = isSpanish ? 'es_MX' : 'en_US';
  // We import and set i18n options within this method, for
  // scenarios where an app needs to import this method and
  // extend the existing list of default links
  const t = translate;
  const i18nOptions = { lng: locale };

  // NOTE: order matters here and links will be displayed in order added to the arrays
  const loggedOut = [];
  const loggedIn = [];

  // Links other than the ones inside this if should need to be explicitly hidden for their
  // respective variations. This means the language and login will show even if a custom set
  // of links is passed in.
  if (!customLinksPassedIn) {
    loggedIn.push({
      label: t('header.myApplicationsAndCoverage', i18nOptions),
      href: `${primaryDomain}/marketplace/auth/global/${ffmLocalePath}/myProfile#landingPage`,
    });
    loggedIn.push({
      label: t('header.myProfile', i18nOptions),
      href: `${primaryDomain}/marketplace/auth/global/${ffmLocalePath}/myProfile#settings`,
    });
  }

  if (!hideLanguageSwitch) {
    const locLink = localeLink(t, locale, subpath, switchLocaleLink);
    loggedOut.push(locLink);
    loggedIn.push(locLink);
  }

  if (!hideLoginLink) {
    const logLink = loginLink(t, deConsumer, primaryDomain);
    loggedOut.push(Object.assign({ identifier: LINK_IDENTIFIERS.LOGIN }, logLink));
  }

  if (!hideLogoutLink) {
    loggedIn.push({
      identifier: LINK_IDENTIFIERS.LOGOUT,
      label: t('header.logout', i18nOptions),
      href: `${primaryDomain || ''}/logout`,
    });
  }

  const links = {};
  links[VARIATION_NAMES.LOGGED_OUT] = loggedOut;
  links[VARIATION_NAMES.LOGGED_IN] = loggedIn;

  return links;
}

export const LINK_IDENTIFIERS = {
  LOGIN: 'login',
  LOGOUT: 'logout',
};

export default defaultMenuLinks;
