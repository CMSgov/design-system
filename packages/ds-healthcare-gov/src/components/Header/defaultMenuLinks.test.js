import defaultMenuLinks, {
  defaultMenuLinks as namedExportDefaultMenuLinks,
} from './defaultMenuLinks';

describe('MenuList', function () {
  it('includes a named export', () => {
    // Some apps may need to import the default menu links in order to extend them
    expect(typeof namedExportDefaultMenuLinks).toBe('function');
  });

  it('leaves out login link if hideLoginLink true', () => {
    expect(
      defaultMenuLinks(undefined, undefined, undefined, undefined, undefined, true)
    ).toMatchSnapshot();
  });

  it('leaves out logout link if hideLogoutLink true', () => {
    expect(
      defaultMenuLinks(undefined, undefined, undefined, undefined, undefined, undefined, true)
    ).toMatchSnapshot();
  });

  it('leaves out locale links if hideLanguageSwitch true', () => {
    expect(
      defaultMenuLinks(
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        true
      )
    ).toMatchSnapshot();
  });

  it('leaves out the myProfile and myApplicationsAndCoverage when customLinksPassedIn is true', () => {
    expect(
      defaultMenuLinks(
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        true
      )
    ).toMatchSnapshot();
  });

  describe('English', () => {
    it('returns array of menu list objects', () => {
      expect(defaultMenuLinks()).toMatchSnapshot();
    });

    it('returns array of menu list objects with subpath', () => {
      expect(defaultMenuLinks('en', 'tax-tool/')).toMatchSnapshot();
    });

    it('returns array of menu list objects with absolute URLs', () => {
      expect(defaultMenuLinks('en', undefined, 'https://www.healthcare.gov')).toMatchSnapshot();
    });

    it('returns array of menu list objects with custom locale switch link', () => {
      expect(
        defaultMenuLinks('en', undefined, undefined, 'https://ayudalocal.cuidadodesalud.gov/es')
      ).toMatchSnapshot();
    });
  });

  describe('Spanish', () => {
    it('returns array of menu list objects', () => {
      expect(defaultMenuLinks('es')).toMatchSnapshot();
    });

    it('returns array of menu list objects with subpath', () => {
      expect(defaultMenuLinks('es', 'tax-tool/')).toMatchSnapshot();
    });

    it('returns array of menu list objects with absolute URLs', () => {
      expect(defaultMenuLinks('es', undefined, 'https://www.healthcare.gov')).toMatchSnapshot();
    });

    it('returns array of menu list objects with custom locale switch link', () => {
      expect(
        defaultMenuLinks('es', undefined, undefined, 'https://localhelp.healthcare.gov')
      ).toMatchSnapshot();
    });
  });
});
