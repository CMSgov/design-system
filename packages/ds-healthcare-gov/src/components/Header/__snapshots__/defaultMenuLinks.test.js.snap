// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuList English returns array of menu list objects 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList English returns array of menu list objects with absolute URLs 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/https://www.healthcare.gov",
      "label": "header.español",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/https://www.healthcare.gov",
      "label": "header.español",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList English returns array of menu list objects with custom locale switch link 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "https://ayudalocal.cuidadodesalud.gov/es/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://ayudalocal.cuidadodesalud.gov/es/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "https://ayudalocal.cuidadodesalud.gov/es/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "https://ayudalocal.cuidadodesalud.gov/es/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList English returns array of menu list objects with subpath 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/login?check_de=1",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList Spanish returns array of menu list objects 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"es\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList Spanish returns array of menu list objects with absolute URLs 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "https://www.healthcare.gov/https://www.healthcare.gov",
      "label": "header.english",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"es\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.healthcare.gov/https://www.healthcare.gov",
      "label": "header.english",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList Spanish returns array of menu list objects with custom locale switch link 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "https://localhelp.healthcare.gov/marketplace/auth/global/es_MX/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "https://localhelp.healthcare.gov/marketplace/auth/global/es_MX/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "https://localhelp.healthcare.gov/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"es\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "https://localhelp.healthcare.gov/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList Spanish returns array of menu list objects with subpath 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/es_MX/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"es\\"}",
    },
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"es\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.healthcare.gov/",
      "label": "header.english",
    },
    Object {
      "href": "/login?check_de=1",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList leaves out locale links if hideLanguageSwitch true 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList leaves out login link if hideLoginLink true 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
  ],
}
`;

exports[`MenuList leaves out logout link if hideLogoutLink true 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#landingPage",
      "label": "header.myApplicationsAndCoverage | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "/marketplace/auth/global/en_US/myProfile#settings",
      "label": "header.myProfile | {\\"lng\\":\\"en\\"}",
    },
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;

exports[`MenuList leaves out the myProfile and myApplicationsAndCoverage when customLinksPassedIn is true 1`] = `
Object {
  "logged-in": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/logout",
      "identifier": "logout",
      "label": "header.logout | {\\"lng\\":\\"en\\"}",
    },
  ],
  "logged-out": Array [
    Object {
      "href": "https://www.cuidadodesalud.gov/es/",
      "label": "header.español",
    },
    Object {
      "href": "/login",
      "identifier": "login",
      "label": "header.login",
    },
  ],
}
`;
